llarp/router/outbound_message_handler.cpp

```diff
@@ -0,0 +1,209 @@
+    }
+
+    if (shouldCreateSession)
+    {
+      QueueSessionCreation(remote);
+    }
+
+    return true;
+  }
+
+  //TODO: this
+  util::StatusObject
+  OutboundMessageHandler::ExtractStatus() const
+  {
+    util::StatusObject status;
+    return status;
+  }
+
+  void
+  Init(LinkManager* linkManager,
+       std::shared_ptr< Logic > logic)
+  {
+    _linkManager = linkManager;
+    _logic = logic;
+  }
+
+  bool
+  OutboundMessageHandler::OnSessionEstablished(const RouterID &router)
+  {
+    FinalizeRequest(router, SendStatus::Success);
+  }
+
+  void
+  OutboundMessageHandler::OnConnectFailure(const RouterID &router)
+  {
+    FinalizeRequest(router, SendStatus::Timeout);
+  }
+
+  void
+  OutboundMessageHandler::OnRouterNotFound(const RouterID &router)
+  {
+    FinalizeRequest(router, SendStatus::RouterNotFound);
+  }
+
+  void
+  OutboundMessageHandler::OnInvalidRouter(const RouterID &router)
+  {
+    FinalizeRequest(router, SendStatus::InvalidRouter);
+  }
+
+  void
+  OutboundMessageHandler::OnNoLink(const RouterID &router)
+  {
+    FinalizeRequest(router, SendStatus::NoLink);
+  }
+
+  void
+  OutboundMessageHandler::OnSessionResult(const RouterID &router,
+                                          const SessionResult result)
+  {
+    switch(result)
+    {
+      case Establish:
+        OnSessionEstablished(router);
+        break;
+      case Timeout:
+        OnConnectFailure(router);
+        break;
+      case RouterNotFound:
+        OnRouterNotFound(router);
+        break;
+      case InvalidRouter:
+        OnInvalidRouter(router);
+        break;
+      case NoLink:
+        OnNoLink(router);
+        break;
+      default:
+        break;
+    }
+  }
+
+  void
+  OutboundMessageHandler::DoCallback(SendStatusHandler callback,
+                                     SendStatus status)
+  {
+    if (callback)
+    {
+      auto func = std::bind(callback, status);
+      _logic->queue_func(func);
+    }
+  }
+
+  void
+  OutboundMessageHandler::QueueSessionCreation(const RouterID& remote)
+  {
+    auto fn = util::memFn(&OutboundMessageHandler::OnSessionResult, this);
+    _linkManager->GetSessionMaker()
+                .CreateSessionTo(remote, fn);
+  }
+
+  bool
+  OutboundMessageHandler::EncodeBuffer(const ILinkMessage *msg,
+                                       llarp_buffer_t& buf)
+  {
+    if(!msg->BEncode(&buf))
+    {
+      LogWarn("failed to encode outbound message, buffer size left: ",
+              buf.size_left());
+      return false;
+    }
+    // set size of message
+    buf.sz  = buf.cur - buf.base;
+    buf.cur = buf.base;
+
+    return true;
+  }
+
+  bool
+  OutboundMessageHandler::Send(const RouterID& remote,
+                               const Message& msg)
+  {
+    llarp_buffer_t buf(msg.first);
+    if (_linkManager->SendTo(remote, buf))
+    {
+      DoCallback(msg.second, SendStatus::Success);
+    }
+    DoCallback(msg.second, SendStatus::Congestion);
```

Comment:
is this what you mean? we always call the congestion func rn

Reply:
I'm not sure what you mean by this, but I don't see any congestion function anywhere in the old `SendToOrQueue` codepath.

Reply:
When I reviewed the function looked like https://github.com/loki-project/loki-network/blob/1d6f010ea9056ff4deee38cd97de3d47727095bb/llarp/router/outbound_message_handler.cpp#L162-L169